internal/controller/apiextensions/usage/reconciler.go

```diff
@@ -0,0 +1,405 @@
+				Kind:       by.Kind,
+				Name:       by.ResourceRef.Name,
+				APIVersion: by.APIVersion,
+			}))
+			// Get the using resource
+			err := r.client.Get(ctx, client.ObjectKey{Name: by.ResourceRef.Name}, using)
+			if xpresource.IgnoreNotFound(err) != nil {
+				log.Debug(errGetUsing, "error", err)
+				err = errors.Wrap(xpresource.IgnoreNotFound(err), errGetUsing)
+				r.record.Event(u, event.Warning(reasonGetUsing, err))
+				return reconcile.Result{}, err
+			}
+
+			if err == nil {
+				// Using resource is still there, so we need to wait for it to be deleted.
+				msg := "Waiting for using resource to be deleted."
+				log.Debug(msg)
+				r.record.Event(u, event.Normal(reasonWaitUsing, msg))
+				return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+			}
+		}
+		// At this point using resource is either:
+		// - not defined
+		// - not found (e.g. deleted)
+		// So, we can proceed with the deletion of the usage.
+
+		// Get the used resource
+		var err error
+		if err = r.client.Get(ctx, client.ObjectKey{Name: of.ResourceRef.Name}, used); xpresource.IgnoreNotFound(err) != nil {
+			log.Debug(errGetUsed, "error", err)
+			err = errors.Wrap(err, errGetUsed)
+			r.record.Event(u, event.Warning(reasonGetUsed, err))
+			return reconcile.Result{}, err
+		}
+
+		// Remove the in-use label from the used resource if no other usages
+		// exists.
+		if err == nil {
+			usageList := &v1alpha1.UsageList{}
+			if err = r.client.List(ctx, usageList, client.MatchingFields{usage.InUseIndexKey: usage.IndexValueForObject(used.GetUnstructured())}); err != nil {
+				log.Debug(errListUsages, "error", err)
+				err = errors.Wrap(err, errListUsages)
+				r.record.Event(u, event.Warning(reasonListUsages, err))
+				return reconcile.Result{}, err
+			}
+			// There are no "other" usageResource's referencing the used resource,
+			// so we can remove the in-use label from the used resource
+			if len(usageList.Items) < 2 {
+				meta.RemoveLabels(used, inUseLabelKey)
+				if err = r.client.Update(ctx, used); err != nil {
+					log.Debug(errRemoveInUseLabel, "error", err)
+					err = errors.Wrap(err, errRemoveInUseLabel)
+					r.record.Event(u, event.Warning(reasonRemoveInUseLabel, err))
+					return reconcile.Result{}, err
+				}
+			}
+		}
+
+		// Remove the finalizer from the usage
+		if err = r.usage.RemoveFinalizer(ctx, u); err != nil {
+			log.Debug(errRemoveFinalizer, "error", err)
+			err = errors.Wrap(err, errRemoveFinalizer)
+			r.record.Event(u, event.Warning(reasonRemoveFinalizer, err))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Add finalizer for Usage resource.
+	if err := r.usage.AddFinalizer(ctx, u); err != nil {
+		log.Debug(errAddFinalizer, "error", err)
+		err = errors.Wrap(err, errAddFinalizer)
+		r.record.Event(u, event.Warning(reasonAddFinalizer, err))
+		return reconcile.Result{}, err
+	}
+
+	d := detailsAnnotation(u)
+	if u.GetAnnotations()[detailsAnnotationKey] != d {
```

Comment:
Just curious - is there a reason to store this as an annotation instead of in a `status` field?  It feels like status to me, but I can be convinced.  Certainly not blocking.

Reply:
This annotation does not add any real value/additional information; the only purpose it is to be able to have a CRD column (i.e. `DETAILS`) with some dynamic value. Otherwise, I would agree with moving to status, but I think annotation is more appropriate because of this. For example:

```bash
❯ kubectl get usages
NAME                 DETAILS                                                     READY   AGE
protect-a-resource   This resource is protected!                                 True    15s
using-uses-used      NopResource/using-resource uses NopResource/used-resource   True    19s
```

In the first one, it is simply `spec.reason`. In the second case, it builds a string again using what is already available in the CR spec, i.e. `fmt.Sprintf("%s/%s uses %s/%s", u.Spec.By.Kind, u.Spec.By.ResourceRef.Name, u.Spec.Of.Kind, u.Spec.Of.ResourceRef.Name)`.